File: server/routers/tasks.py
"""Task routes — CRUD with role-based access control."""

from uuid import UUID
from typing import Optional, Union, Dict, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from server.database import get_db
from server.models import Task, User, Project, ProjectMember
from server.schemas import TaskCreate, TaskUpdate
from server.auth import get_current_user, require_project_role

router = APIRouter(
    prefix="/api/projects/{project_id}/tasks", tags=["tasks"]
)


# ── helpers ──────────────────────────────────────

def _serialize_user(u: Optional[User]) -> Optional[Dict[str, Any]]:
    if u is None:
        return None
    return {"id": str(u.id), "name": u.name, "email": u.email}


def _serialize_task(t: Task) -> dict:
    return {
        "id": str(t.id),
        "project_id": str(t.project_id),
        "title": t.title,
        "description": t.description or "",
        "status": t.status,
        "priority": t.priority,
        "due_date": t.due_date.isoformat() if t.due_date else None,
        "assigned_to": str(t.assigned_to) if t.assigned_to else None,
        "created_by": str(t.created_by),
        "created_at": t.created_at.isoformat() if t.created_at else None,
        "updated_at": t.updated_at.isoformat() if t.updated_at else None,
        "assignee": _serialize_user(t.assignee),
        "creator": _serialize_user(t.creator),
    }


async def _load_task(db, task_id):
    result = await db.execute(
        select(Task)
        .options(
            selectinload(Task.assignee),
            selectinload(Task.creator),
        )
        .where(Task.id == task_id)
    )
    return result.scalar_one_or_none()


# ── GET /api/projects/{project_id}/tasks ─────────
@router.get("")
async def list_tasks(
    project_id: UUID,
    status: Optional[str] = Query(None),
    priority: Optional[str] = Query(None),
    assigned_to: Optional[UUID] = Query(None),
    membership: ProjectMember = Depends(require_project_role("admin", "member")),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    conditions = [Task.project_id == project_id]

    # Members can only see tasks assigned to themselves
    if membership.role == "member":
        conditions.append(Task.assigned_to == current_user.id)

    # Optional filters
    if status:
        if status not in ("todo", "in_progress", "done"):
            raise HTTPException(
                status_code=400, detail="Status must be todo, in_progress, or done"
            )
        conditions.append(Task.status == status)

    if priority:
        if priority not in ("low", "medium", "high"):
            raise HTTPException(
                status_code=400, detail="Priority must be low, medium, or high"
            )
        conditions.append(Task.priority == priority)

    if assigned_to:
        if membership.role == "member" and assigned_to != current_user.id:
            raise HTTPException(
                status_code=403,
                detail="Members can only filter tasks assigned to themselves",
            )
        conditions.append(Task.assigned_to == assigned_to)

    result = await db.execute(
        select(Task)
        .options(selectinload(Task.assignee), selectinload(Task.creator))
        .where(*conditions)
        .order_by(Task.priority.asc(), Task.created_at.desc())
    )
    tasks = result.scalars().all()

    return {"tasks": [_serialize_task(t) for t in tasks]}


# ── GET /api/projects/{project_id}/tasks/{task_id} ──
@router.get("/{task_id}")
async def get_task(
    project_id: UUID,
    task_id: UUID,
    membership: ProjectMember = Depends(require_project_role("admin", "member")),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    conditions = [Task.id == task_id, Task.project_id == project_id]
    if membership.role == "member":
        conditions.append(Task.assigned_to == current_user.id)

    result = await db.execute(
        select(Task)
        .options(
            selectinload(Task.assignee),
            selectinload(Task.creator),
            selectinload(Task.project),
        )
        .where(*conditions)
    )
    task = result.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    d = _serialize_task(task)
    d["project"] = {"id": str(task.project.id), "name": task.project.name} if task.project else None
    return {"task": d}


# ── POST /api/projects/{project_id}/tasks ────────
@router.post("", status_code=201)
async def create_task(
    project_id: UUID,
    body: TaskCreate,
    membership: ProjectMember = Depends(require_project_role("admin")),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Validate assigned_to is a project member
    if body.assigned_to:
        result = await db.execute(
            select(ProjectMember).where(
                ProjectMember.project_id == project_id,
                ProjectMember.user_id == body.assigned_to,
            )
        )
        if not result.scalar_one_or_none():
            raise HTTPException(
                status_code=400,
                detail="Assigned user is not a member of this project",
            )

    task = Task(
        project_id=project_id,
        title=body.title.strip(),
        description=(body.description or "").strip(),
        status=body.status or "todo",
        priority=body.priority or "medium",
        due_date=body.due_date,
        assigned_to=body.assigned_to,
        created_by=current_user.id,
    )
    db.add(task)
    await db.commit()

    full = await _load_task(db, task.id)
    return {"task": _serialize_task(full)}


# ── PUT /api/projects/{project_id}/tasks/{task_id} ──
@router.put("/{task_id}")
async def update_task(
    project_id: UUID,
    task_id: UUID,
    body: TaskUpdate,
    membership: ProjectMember = Depends(require_project_role("admin", "member")),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Task).where(Task.id == task_id, Task.project_id == project_id)
    )
    task = result.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    if membership.role == "member":
        # Members can only update status on their own tasks
        if task.assigned_to != current_user.id:
            raise HTTPException(
                status_code=403,
                detail="You can only update tasks assigned to you",
            )

        update_data = body.model_dump(exclude_unset=True)
        allowed_fields = {"status"}
        disallowed = set(update_data.keys()) - allowed_fields
        if disallowed:
            raise HTTPException(
                status_code=403,
                detail=f"Members can only update task status. Cannot modify: {', '.join(disallowed)}",
            )

        if body.status is not None:
            task.status = body.status
    else:
        # Admin: can update everything
        if body.title is not None:
            task.title = body.title.strip()
        if body.description is not None:
            task.description = body.description.strip()
        if body.status is not None:
            task.status = body.status
        if body.priority is not None:
            task.priority = body.priority
        if body.due_date is not None:
            task.due_date = body.due_date

        update_data = body.model_dump(exclude_unset=True)
        if "assigned_to" in update_data:
            new_assignee = body.assigned_to
            if new_assignee:
                result2 = await db.execute(
                    select(ProjectMember).where(
                        ProjectMember.project_id == project_id,
                        ProjectMember.user_id == new_assignee,
                    )
                )
                if not result2.scalar_one_or_none():
                    raise HTTPException(
                        status_code=400,
                        detail="Assigned user is not a member of this project",
                    )
            task.assigned_to = new_assignee

    await db.commit()

    full = await _load_task(db, task.id)
    return {"task": _serialize_task(full)}


# ── DELETE /api/projects/{project_id}/tasks/{task_id} ──
@router.delete("/{task_id}")
async def delete_task(
    project_id: UUID,
    task_id: UUID,
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Task).where(Task.id == task_id, Task.project_id == project_id)
    )
    task = result.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    await db.delete(task)
    await db.commit()

    return {"message": "Task deleted successfully"}

File: server/routes/tasks.js
const express = require('express');
const { body, param, query, validationResult } = require('express-validator');
const { Op } = require('sequelize');
const { Task, User, Project } = require('../models');
const auth = require('../middleware/auth');
const roleCheck = require('../middleware/roleCheck');

const router = express.Router({ mergeParams: true });

// All task routes require authentication
router.use(auth);

// ──────────────────────────────────────────────
// GET /api/projects/:projectId/tasks — List tasks with optional filters
// ──────────────────────────────────────────────
router.get(
  '/',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    query('status')
      .optional()
      .isIn(['todo', 'in_progress', 'done'])
      .withMessage('Status must be todo, in_progress, or done'),
    query('priority')
      .optional()
      .isIn(['low', 'medium', 'high'])
      .withMessage('Priority must be low, medium, or high'),
    query('assigned_to')
      .optional()
      .isUUID()
      .withMessage('assigned_to must be a valid UUID'),
  ],
  roleCheck('admin', 'member'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const where = { project_id: req.params.projectId };
      const userRole = req.membership.role;

      // Members should only see tasks assigned to themselves.
      if (userRole === 'member') {
        where.assigned_to = req.user.id;
      }

      // Apply optional filters
      if (req.query.status) where.status = req.query.status;
      if (req.query.priority) where.priority = req.query.priority;
      if (req.query.assigned_to) {
        // Members cannot override the self-assigned visibility filter.
        if (userRole === 'member' && req.query.assigned_to !== req.user.id) {
          return res.status(403).json({
            message: 'Members can only filter tasks assigned to themselves',
          });
        }
        where.assigned_to = req.query.assigned_to;
      }

      const tasks = await Task.findAll({
        where,
        include: [
          { model: User, as: 'assignee', attributes: ['id', 'name', 'email'] },
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
        ],
        order: [
          ['priority', 'ASC'], // high first when ordered by enum: high=0 contextually, but stored as string
          ['created_at', 'DESC'],
        ],
      });

      res.json({ tasks });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// GET /api/projects/:projectId/tasks/:taskId — Get single task
// ──────────────────────────────────────────────
router.get(
  '/:taskId',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    param('taskId').isUUID().withMessage('Invalid task ID'),
  ],
  roleCheck('admin', 'member'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const taskWhere = { id: req.params.taskId, project_id: req.params.projectId };
      if (req.membership.role === 'member') {
        taskWhere.assigned_to = req.user.id;
      }

      const task = await Task.findOne({
        where: taskWhere,
        include: [
          { model: User, as: 'assignee', attributes: ['id', 'name', 'email'] },
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
          { model: Project, as: 'project', attributes: ['id', 'name'] },
        ],
      });

      if (!task) {
        return res.status(404).json({ message: 'Task not found' });
      }

      res.json({ task });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// POST /api/projects/:projectId/tasks — Create task (admin only)
// ──────────────────────────────────────────────
router.post(
  '/',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    body('title')
      .trim()
      .notEmpty()
      .withMessage('Task title is required')
      .isLength({ max: 255 })
      .withMessage('Task title must be at most 255 characters'),
    body('description')
      .optional()
      .trim(),
    body('status')
      .optional()
      .isIn(['todo', 'in_progress', 'done'])
      .withMessage('Status must be todo, in_progress, or done'),
    body('priority')
      .optional()
      .isIn(['low', 'medium', 'high'])
      .withMessage('Priority must be low, medium, or high'),
    body('due_date')
      .optional({ values: 'null' })
      .isISO8601()
      .withMessage('Due date must be a valid date (YYYY-MM-DD)'),
    body('assigned_to')
      .optional({ values: 'null' })
      .isUUID()
      .withMessage('assigned_to must be a valid UUID'),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const { title, description, status, priority, due_date, assigned_to } = req.body;

      // If assigned_to is provided, verify the user is a member of the project
      if (assigned_to) {
        const { ProjectMember } = require('../models');
        const isMember = await ProjectMember.findOne({
          where: { project_id: req.params.projectId, user_id: assigned_to },
        });
        if (!isMember) {
          return res.status(400).json({ message: 'Assigned user is not a member of this project' });
        }
      }

      const task = await Task.create({
        project_id: req.params.projectId,
        title,
        description: description || '',
        status: status || 'todo',
        priority: priority || 'medium',
        due_date: due_date || null,
        assigned_to: assigned_to || null,
        created_by: req.user.id,
      });

      const fullTask = await Task.findByPk(task.id, {
        include: [
          { model: User, as: 'assignee', attributes: ['id', 'name', 'email'] },
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
        ],
      });

      res.status(201).json({ task: fullTask });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// PUT /api/projects/:projectId/tasks/:taskId — Update task
// Admin: can update everything
// Member: can only update status of tasks assigned to them
// ──────────────────────────────────────────────
router.put(
  '/:taskId',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    param('taskId').isUUID().withMessage('Invalid task ID'),
    body('title')
      .optional()
      .trim()
      .notEmpty()
      .withMessage('Task title cannot be empty')
      .isLength({ max: 255 })
      .withMessage('Task title must be at most 255 characters'),
    body('description')
      .optional()
      .trim(),
    body('status')
      .optional()
      .isIn(['todo', 'in_progress', 'done'])
      .withMessage('Status must be todo, in_progress, or done'),
    body('priority')
      .optional()
      .isIn(['low', 'medium', 'high'])
      .withMessage('Priority must be low, medium, or high'),
    body('due_date')
      .optional({ values: 'null' })
      .isISO8601()
      .withMessage('Due date must be a valid date (YYYY-MM-DD)'),
    body('assigned_to')
      .optional({ values: 'null' })
      .isUUID()
      .withMessage('assigned_to must be a valid UUID'),
  ],
  roleCheck('admin', 'member'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const task = await Task.findOne({
        where: { id: req.params.taskId, project_id: req.params.projectId },
      });

      if (!task) {
        return res.status(404).json({ message: 'Task not found' });
      }

      const userRole = req.membership.role;

      // Member can only update status of tasks assigned to them
      if (userRole === 'member') {
        if (task.assigned_to !== req.user.id) {
          return res.status(403).json({ message: 'You can only update tasks assigned to you' });
        }

        const allowedFields = ['status'];
        const attemptedFields = Object.keys(req.body);
        const disallowedFields = attemptedFields.filter((f) => !allowedFields.includes(f));
        if (disallowedFields.length > 0) {
          return res.status(403).json({
            message: `Members can only update task status. Cannot modify: ${disallowedFields.join(', ')}`,
          });
        }

        if (req.body.status) {
          task.status = req.body.status;
        }
      } else {
        // Admin: can update everything
        const { title, description, status, priority, due_date, assigned_to } = req.body;

        if (title !== undefined) task.title = title;
        if (description !== undefined) task.description = description;
        if (status !== undefined) task.status = status;
        if (priority !== undefined) task.priority = priority;
        if (due_date !== undefined) task.due_date = due_date || null;

        if (assigned_to !== undefined) {
          if (assigned_to) {
            const { ProjectMember } = require('../models');
            const isMember = await ProjectMember.findOne({
              where: { project_id: req.params.projectId, user_id: assigned_to },
            });
            if (!isMember) {
              return res.status(400).json({ message: 'Assigned user is not a member of this project' });
            }
          }
          task.assigned_to = assigned_to || null;
        }
      }

      await task.save();

      const updatedTask = await Task.findByPk(task.id, {
        include: [
          { model: User, as: 'assignee', attributes: ['id', 'name', 'email'] },
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
        ],
      });

      res.json({ task: updatedTask });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// DELETE /api/projects/:projectId/tasks/:taskId — Delete task (admin only)
// ──────────────────────────────────────────────
router.delete(
  '/:taskId',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    param('taskId').isUUID().withMessage('Invalid task ID'),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const task = await Task.findOne({
        where: { id: req.params.taskId, project_id: req.params.projectId },
      });

      if (!task) {
        return res.status(404).json({ message: 'Task not found' });
      }

      await task.destroy();

      res.json({ message: 'Task deleted successfully' });
    } catch (error) {
      next(error);
    }
  }
);

module.exports = router;
